src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Date.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json
+{
+    internal static partial class JsonWriterHelper
+    {
+        public static bool TryFormatToISO(DateTimeOffset value, Span<byte> destination, out int bytesWritten)
+        {
+            return TryFormatDateTime(value.DateTime, value.Offset, destination, out bytesWritten);
+        }
+
+        public static bool TryFormatToISO(DateTime value, Span<byte> destination, out int bytesWritten)
+        {
+            return TryFormatDateTime(value, JsonConstants.NullUtcOffset, destination, out bytesWritten);
+        }
+
+        //
+        // ISO 8601 format
+        // If the milliseconds part of the date is zero, we omit the fraction part of the output string,
+        // else we write the fraction up to 7 decimal places with no trailing zeros. i.e. the format is
+        // YYYY-MM-DDThh:mm:ss[.s]TZD where TZD = Z or +-hh:mm.
+        // e.g.
+        //   ---------------------------------
+        //   2017-06-12T05:30:45.768-07:00
+        //   2017-06-12T05:30:45.00768Z           (Z is short for "+00:00" but also distinguishes DateTimeKind.Utc from DateTimeKind.Local)
+        //   2017-06-12T05:30:45                  (interpreted as local time wrt to current time zone)
+        private static bool TryFormatDateTime(DateTime value, TimeSpan offset, Span<byte> destination, out int bytesWritten)
+        {
+            // Must write YYYY-MM-DDThh:mm:ss
+            const int MinimumBytesNeeded = 19;
+
+            int bytesRequired = MinimumBytesNeeded;
+            DateTimeKind kind = DateTimeKind.Local;
+
+            if (offset == JsonConstants.NullUtcOffset)
+            {
+                kind = value.Kind;
+                if (kind == DateTimeKind.Local)
+                {
+                    offset = TimeZoneInfo.Local.GetUtcOffset(value);
+                    bytesRequired += 6;
+                }
+                else if (kind == DateTimeKind.Utc)
+                {
+                    bytesRequired += 1;
+                }
+            }
+            else
+            {
+                bytesRequired += 6;
+            }
+
+            if (destination.Length < bytesRequired)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            bytesWritten = bytesRequired;
+
+            uint fraction = (uint)(value.Ticks % TimeSpan.TicksPerSecond);
+            int numFractionDigits = 0;
+
+            if (fraction > 0)
```

Comment:
This fractional component part should be moved down after writing the `seconds` value. We have `if (fraction > 0)` branch there as well.